helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -29,6 +31,8 @@
  */
 abstract class HardConstraint {
 
+  protected Set<String> logEnabledReplicas;
```

Comment:
As I mentioned, better have a flag for constraint level. Replica level creates a lot of string objects for that. That's too much.

Reply:
Yes, Updated the PR.